disdat/utility/aws_s3.py

```diff
@@ -374,30 +396,22 @@ def ls_s3_url_keys(s3_url, is_object_directory=False):
 
     hexes = '0123456789abcdef'
 
-    multiple_results = []
-
-    # MacOS X fails when we multi-process using fork and boto sessions.
-    # One fix is to set export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
-    mp_ctxt = get_context(MP_CONTEXT_TYPE)
-
-    est_cpu_count = mp_ctxt.cpu_count()
-    print("ls_s3_url_keys using MP with cpu_count {}".format(est_cpu_count))
-    pool = mp_ctxt.Pool(processes=est_cpu_count, maxtasksperchild=MAX_TASKS_PER_CHILD)
-
-    prefixes = ['']
-    if is_object_directory:
-        prefixes = [f'{c}{d}' for c in hexes for d in hexes]
-
     results = []
 
     if is_object_directory:
-        for i, prefix in enumerate(prefixes):
-            prefix = os.path.join(s3_path, prefix)
-            multiple_results.append(pool.apply_async(s3_list_objects_at_prefix_v2,
-                                                     (bucket, prefix),
-                                                     callback=results.extend))
-        pool.close()
-        pool.join()
+        prefixes = [f'{c}{d}' for c in hexes for d in hexes]
+        multiple_results = []
+        mp_ctxt = get_context(MP_CONTEXT_TYPE)
+        est_cpu_count = disdat_cpu_count()
+        print("ls_s3_url_keys using MP with cpu_count {}".format(est_cpu_count))
+        with mp_ctxt.Pool(processes=est_cpu_count, maxtasksperchild=MAX_TASKS_PER_CHILD) as pool:
+            for i, prefix in enumerate(prefixes):
+                prefix = os.path.join(s3_path, prefix)
+                multiple_results.append(pool.apply_async(s3_list_objects_at_prefix_v2,
+                                                         (bucket, prefix),
+                                                         callback=results.extend))
+            pool.close()
```

Comment:
do you need a close() with the context manager?

Reply:
pool requires close before join, and I'd still like only exit this call when all the tasks are complete.